### PR TITLE
[#173035929] code generation + bonus client

### PR DESCRIPTION
**Pull request by balanza**
This is a draft as we need to wait for [this](https://github.com/pagopa/io-functions-bonus/pull/6) to be merged first.

* add generation script in package.json
* create a barebone client

### CODE REVIEW

File: src/clients/bonus.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  StartBonusEligibilityCheckT
+} from "../../generated/io-bonus-api/requestTypes";
+
+// we want to authenticate against the platform APIs with the APIM header key or
+// the Azure Functions header key, so we send both headers
+function SubscriptionKeyHeaderProducer<P>(
+  token: string
+): RequestHeaderProducer<P, "Ocp-Apim-Subscription-Key"> {
+  return () => ({
+    "Ocp-Apim-Subscription-Key": token,
```
Comment (gunzip) [inline src/clients/bonus.ts:24]:
```suggestion
```

File: src/clients/bonus.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import { Omit } from "italia-ts-commons/lib/types";
+import nodeFetch from "node-fetch";
+
+import {
+  startBonusEligibilityCheckDefaultDecoder,
+  StartBonusEligibilityCheckT
+} from "../../generated/io-bonus-api/requestTypes";
+
+// we want to authenticate against the platform APIs with the APIM header key or
+// the Azure Functions header key, so we send both headers
```
Comment (gunzip) [inline src/clients/bonus.ts:19]:
```suggestion
// we want to authenticate against the platform APIs with
// the Azure Functions header key
```

File: src/clients/bonus.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  const startBonusEligibilityCheckT: ReplaceRequestParams<
+    StartBonusEligibilityCheckT,
+    Omit<RequestParams<StartBonusEligibilityCheckT>, "SubscriptionKey">
+  > = {
+    body: _ => "",
+    headers: composeHeaderProducers(tokenHeaderProducer, ApiHeaderJson),
+    method: "post",
+    query: _ => ({}),
+    response_decoder: startBonusEligibilityCheckDefaultDecoder(),
+    url: params => `/profiles/${params.fiscalCode}`
```
Comment (gunzip) [inline src/clients/bonus.ts:54]: need fix

Comment (gunzip) [reply src/clients/bonus.ts:54]: why /`profile` ? isn't this `/bonus/vacanze/eligibility` ?

File: src/clients/bonus.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+function SubscriptionKeyHeaderProducer<P>(
+  token: string
+): RequestHeaderProducer<P, "Ocp-Apim-Subscription-Key"> {
+  return () => ({
+    "Ocp-Apim-Subscription-Key": token,
+    "X-Functions-Key": token
+  });
+}
+
+export function APIClient(
```
Comment (BurnedMarshal) [inline src/clients/bonus.ts:29]: we can call it `BonusAPIClient`? To avoid naming conflict with the already existing `APIClient`

File: src/services/bonusApiClientFactory.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
+/**
+ * This service builds Bonus API clients.
+ */
+
+import nodeFetch from "node-fetch";
+
+import { BonusAPIClient } from "../clients/bonus";
+import { IBonusAPIClientFactoryInterface } from "./IBonusApiClientFactory";
+
+// TODO: this class is actually useless as BonusAPIClient is immutable, it can be removed
```
Comment (gunzip) [inline src/services/bonusApiClientFactory.ts:10]: exactly, get rid of this class / factory which is useless. we'll call new BonusApiClient() when needed

Comment (balanza) [reply src/services/bonusApiClientFactory.ts:10]: should we do the same for `ApiClientFactory` and `PagoPAClientFactory`? They are actually the same (except pagopa creates a different client based on the environment, but such decision can be made in `config.ts` imho)

Comment (gunzip) [reply src/services/bonusApiClientFactory.ts]: yes but not with this PR :) very low priority

**Comment by gunzip** [comment]
needs rebase and fix linting

File: src/config.ts [hunk 6]
```diff
@@ -101,16 +102,16 @@ const CIE_METADATA_URL = getRequiredENVVar("CIE_METADATA_URL");
...
+          STRINGS_RECORD.decode(_1).mapLeft(
+            err => new Error(errorsToReadableMessages(err).join(" / "))
+          )
         )
-      )
-      .getOrElse(undefined)
-  )
-  .getOrElse(undefined);
+        .getOrElse(undefined)
+    )
```
Comment (gunzip) [inline src/config.ts:113]: it looks like you have an odd configuration of prettier, can you revert these changes?

File: src/services/bonusApiClientFactory.ts (hunk 5 above)
Comment (balanza) [reply src/services/bonusApiClientFactory.ts]: tracked it [here](https://www.pivotaltracker.com/story/show/173075418)
